### ISSUE TITLE
Add a Model.fit-like API for seq2seq models

### ISSUE THREAD

**Issue by luozhouyang** [tensorflow/tensorflow#36875]
Many models can be trained by `tf.keras.Model.fit` API, which is very convenient. But for seq2seq models, we have to use a custom training loop. As we know, seq2seq Models are used very often, so why not build a API to train these models in a convenient way?

**Comment by pandrey-fr** [comment]
Hi,
Could you please elaborate on this statement, and maybe provide a small example?
I personnally use custom sequence-to-sequence models and have no problem whatsoever wrapping them up in keras and using the built-in training loop, so at this point I do not quite understand your issue, but surely I am missing something.

**Comment by luozhouyang** [comment]
Sorry, I didn't make it clear. I mean, I have to use a loop to generate the whole sequence during the prediction in seq2seq models.
Here is an example from [Transformer](https://www.tensorflow.org/tutorials/text/transformer#evaluate):

```python
def evaluate(inp_sentence):
  start_token = [tokenizer_pt.vocab_size]
  end_token = [tokenizer_pt.vocab_size + 1]
  
  # inp sentence is portuguese, hence adding the start and end token
  inp_sentence = start_token + tokenizer_pt.encode(inp_sentence) + end_token
  encoder_input = tf.expand_dims(inp_sentence, 0)
  
  # as the target is english, the first word to the transformer should be the
  # english start token.
  decoder_input = [tokenizer_en.vocab_size]
  output = tf.expand_dims(decoder_input, 0)
    
  for i in range(MAX_LENGTH):
    enc_padding_mask, combined_mask, dec_padding_mask = create_masks(
        encoder_input, output)
  
    # predictions.shape == (batch_size, seq_len, vocab_size)
    predictions, attention_weights = transformer(encoder_input, 
                                                 output,
                                                 False,
                                                 enc_padding_mask,
                                                 combined_mask,
                                                 dec_padding_mask)
    
    # select the last word from the seq_len dimension
    predictions = predictions[: ,-1:, :]  # (batch_size, 1, vocab_size)

    predicted_id = tf.cast(tf.argmax(predictions, axis=-1), tf.int32)
    
    # return the result if the predicted_id is equal to the end token
    if predicted_id == tokenizer_en.vocab_size+1:
      return tf.squeeze(output, axis=0), attention_weights
    
    # concatentate the predicted_id to the output which is given to the decoder
    # as its input.
    output = tf.concat([output, predicted_id], axis=-1)

  return tf.squeeze(output, axis=0), attention_weights
```

Obviously, this process can not be finished by `Model.predict` API.

What i want is, once i create my seq2seq model, I can use something like `predict_loop` method to finish this process.

```python

class Seq2Seq(tf.keras.Model):
    pass

model = Seq2Seq()
# training ... 

generated_sequence = model.predict_loop(inputs, callbacks,...)

```

**Comment by pandrey-fr** [comment]
Oh, I see! So, your problem is indeed with the prediction loop rather than the training one?

In one of my personal working cases, I implemented a custom looping prediction method for the Layer that wraps my model's architecture, but having a general mechanism available in TF would be great. Let's wait for a tensorflower to pick up this issue and give their opinion and expertise on the topic then :)

**Comment by pavithrasv** [comment]
Adding @qlzh727 who is a seq2seq model expert to advice on this. Thank you!

**Comment by fchollet** [comment]
Consider creating a subclass of `Model` that overrides the training step (see method `_train_step`) or the `fit` loop. This is now very easy to do. You'll be able to use such a `Model` subclass to create new models either via the Functional API or via subclassing.

**Comment by pandrey-fr** [comment]
@fchollet Thank you for the tip. I was personnally unaware of the renewed API of `Model`, and it feels amazing. To what extent may we (i.e. individual users) consider `_train_step` a stable API feature? I have been using a homemade gradient stacking framework through a keras Model wrapper for a few months, but re-implementing its core mechanism as a specific `Model` subclass using `_train_step` would be so much more elegant (and would make it possible to suggest it through a PR submission).

**Comment by luozhouyang** [comment]
@fchollet @pavithrasv thanks, I will have a try.
